llvm/tools/ir2builder/ir2builder.cpp

```diff
@@ -0,0 +1,1702 @@
+  case Instruction::PtrToInt: {
+    call = "CreatePtrToInt(" + op1 + ", " + getType(I->getType()) + ")";
+  } break;
+  case Instruction::IntToPtr: {
+    call = "CreateIntToPtr(" + op1 + ", " + getType(I->getType()) + ")";
+  } break;
+  case Instruction::BitCast: {
+    call = "CreateBitCast(" + op1 + ", " + getType(I->getType()) + ")";
+  } break;
+  case Instruction::AddrSpaceCast: {
+    call = "CreateAddrSpaceCast(" + op1 + ", " + getType(I->getType()) + ")";
+  } break;
+  case Instruction::CleanupPad: {
+    auto cupI = dyn_cast<CleanupPadInst>(I);
+    std::string argsStr = "{";
+    bool first = true;
+    for (unsigned i = 0; i < cupI->arg_size(); ++i) {
+      if (!first)
+        argsStr += ", ";
+      argsStr += getVal(cupI->getArgOperand(i));
+      first = false;
+    }
+    argsStr += "}";
+    call = "CreateCleanupPad(" + op1 + ", " + argsStr + ")";
+  } break;
+  case Instruction::CatchPad: {
+    auto capI = dyn_cast<CatchPadInst>(I);
+    std::string argsStr = "{";
+    bool first = true;
+    for (unsigned i = 0; i < capI->arg_size(); ++i) {
+      if (!first)
+        argsStr += ", ";
+      argsStr += getVal(capI->getArgOperand(i));
+      first = false;
+    }
+    argsStr += "}";
+    call = "CreateCatchPad(" + op1 + ", " + argsStr + ")";
+  } break;
+  case Instruction::ICmp: {
+    auto cmpI = dyn_cast<CmpInst>(I);
+    std::string cmpPredicate =
+        llvmPrefix + getCmpPredicate(cmpI->getPredicate());
+    call = "CreateICmp(" + cmpPredicate + ", " + op1 + ", " + op2 + ")";
+  } break;
+  case Instruction::FCmp: {
+    auto cmpI = dyn_cast<CmpInst>(I);
+    std::string cmpPredicate =
+        llvmPrefix + getCmpPredicate(cmpI->getPredicate());
+    call = "CreateFCmp(" + cmpPredicate + ", " + op1 + ", " + op2 + ")";
+  } break;
+  case Instruction::PHI: {
+    auto phI = dyn_cast<PHINode>(I);
+    call = "CreatePHI(" + getType(I->getType()) + ", " +
+           std::to_string(phI->getNumIncomingValues()) + ")";
+    std::string phVar = getVal(phI);
+    OS << "auto " << phVar << " = " << builder << "." << call << ";\n";
+
+    unsigned i = 0;
+    for (auto b : phI->blocks()) {
+      auto incVal = getVal(phI->getIncomingValue(i));
+      // Phis might contain incoming not yet defined variable in such case
+      // we save the line we would output here and output it after the
+      // whole function body was outputted
+      std::string line =
+          phVar + "->addIncoming(" + incVal + ", " + getVal(b) + ");\n";
+      phiIncomings.push_back(line);
+      ++i;
+    }
+    return;
+  } break;
+  case Instruction::Call: {
+    auto fcI = dyn_cast<CallBase>(I);
+    std::string argDecl = "";
+    if (fcI->arg_size() > 0) {
+      argDecl = getNextVar();
+      OS << "Value *" << argDecl << "[] = {";
+      bool isFirst = true;
+      for (unsigned i = 0; i < fcI->arg_size(); ++i) {
+        if (!isFirst)
+          OS << ", ";
+        OS << getVal(fcI->getArgOperand(i));
+        isFirst = false;
+      }
```

Comment:
The idiom from 1515 to 1521 is printing a sequence of values with the "right" number of delimiters (the same idiom  on different types appears a number of times elsewhere). I think `formatv` may be better here too. For example, for a simple vector of integers, `formatv` will do this:
```
std::vector<int> V = {8, 9, 10};
S = formatv("{0}", make_range(V.begin(), V.end())); // S == "8, 9, 10"
```
The difference from the example is that you need to apply `getVal()` to each element first. Perhaps something like this:
```
      SmallVector<std::string> Tmp;
      OS << formatv("{0}", (llvm::transform(fcI->args(), back_inserter(Tmp),
                                            [&](Value *V) { return getVal(V); }),
                            make_range(Tmp.begin(), Tmp.end())));
``` 
That might be further simplified further by writing a simple `format_provider`/`FormatAdapter` (https://llvm.org/docs/ProgrammersManual.html#custom-formatting).